File: LogicAndTrick.WikiCodeParser/Tags/Tag.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using LogicAndTrick.WikiCodeParser.Nodes;

namespace LogicAndTrick.WikiCodeParser.Tags
{
    public class Tag
    {
        public string Token { get; set; }
        public string Element { get; set; }
        public string ElementClass { get; set; }
        public string MainOption { get; set; }
        public string[] Options { get; set; }
        public bool AllOptionsInMain { get; set; }
        public bool IsBlock { get; set; }
        public bool IsNested { get; set; }

        public List<string> Scopes { get; set; }
        public int Priority { get; set; } = 0;

        protected Tag()
        {
            Options = new string[0];
            Scopes = new List<string>();
        }

        public Tag(string token, string element, string elementClass = null) : this()
        {
            Token = token;
            Element = element;
            ElementClass = elementClass;
        }

        public virtual bool InScope(string scope) => string.IsNullOrWhiteSpace(scope) ||
                                                     Scopes.Contains(scope, StringComparer.InvariantCultureIgnoreCase);

        public virtual bool Matches(State state, string token) => token?.ToLower() == Token;

        public virtual INode Parse(Parser parser, ParseData data, State state, string scope)
        {
            var index = state.Index;
            var tokenLength = Token.Length;

            state.Seek(tokenLength + 1, false);
            var optionsString = state.ScanTo("]").Trim();
            if (state.Next() != ']')
            {
                state.Seek(index, true);
                return null;
            }

            var options = new Dictionary<string, string>();
            if (optionsString.Length > 0)
            {
                if (optionsString[0] == '=' && AllOptionsInMain && MainOption != null)
                {
                    options[MainOption] = optionsString.Substring(1);
                }
                else
                {
                    if (optionsString[0] == '=') optionsString = MainOption + optionsString;
                    var matches = Regex.Matches(optionsString, @"(?=\s|^)\s*([^ ]+?)=([^\s]*)\b(?!=)",
                        RegexOptions.IgnoreCase | RegexOptions.Singleline | RegexOptions.Multiline);
                    foreach (Match m in matches)
                    {
                        var name = m.Groups[1].Value.Trim();
                        var value = m.Groups[2].Value.Trim();
                        options[name] = value;
                    }
                }
            }

            if (IsNested)
            {
                var stack = 1;
                var text = "";
                while (!state.Done)
                {
                    text += state.ScanTo("[");
                    var tok = state.GetToken();
                    if (tok == Token) stack++;
                    if (tok == "/" + Token && state.Peek(tokenLength + 3) == "[/" + Token + "]") stack--;
                    if (stack == 0)
                    {
                        state.Seek(Token.Length + 3, false);
                        if (!Validate(options, text)) break;
                        return FormatResult(parser, data, state, scope, options, text);
                    }

                    text += state.Next();
                }

                state.Seek(index, true);
                return null;
            }
            else
            {
                var text = state.ScanTo("[/" + Token + "]");
                if (state.Peek(tokenLength + 3).ToLower() == "[/" + Token + "]" && Validate(options, text))
                {
                    state.Seek(Token.Length + 3, false);
                    return FormatResult(parser, data, state, scope, options, text);
                }
                else
                {
                    state.Seek(index, true);
                    return null;
                }
            }
        }

        public virtual bool Validate(Dictionary<string, string> options, string text) => true;

        public virtual INode FormatResult(Parser parser, ParseData data, State state, string scope, Dictionary<string, string> options, string text)
        {
            var before = "<" + Element;
            if (ElementClass != null) before += " class=\"" + ElementClass + '"';
            before += '>';
            var after = "</" + Element + '>';
            var content = parser.ParseTags(data, text, scope, IsBlock ? "block" : "inline");
            return new HtmlNode(before, content, after);
        }
    }
}

File: javascript/src/Tags/Tag.ts
import { HtmlNode } from '../Nodes/HtmlNode';
import { INode } from '../Nodes/INode';
import { ParseData } from '../ParseData';
import { Parser } from '../Parser';
import { State } from '../State';
import { TagParseContext } from '../TagParseContext';

export class Tag {
    public Token: string;
    public Element: string;
    public ElementClass: string | null;
    public MainOption: string;
    public Options: string[];
    public AllOptionsInMain: boolean;
    public IsBlock: boolean;
    public IsNested: boolean;

    public Scopes: string[];
    public Priority = 0;

    protected TagContext(): TagParseContext {
        return this.IsBlock ? TagParseContext.Block : TagParseContext.Inline;
    }

    constructor(token = '', element = '', elementClass: string | null = null) {
        this.Token = token;
        this.Element = element;
        this.ElementClass = elementClass;
        this.Options = [];
        this.Scopes = [];
    }

    public InScope(scope: string): boolean {
        return !scope
            || scope.trim() == ''
            || this.Scopes.includes(scope);
    }

    public Matches(state: State, token: string, context: TagParseContext): boolean {
        return token?.toLowerCase() == this.Token && (context == TagParseContext.Block || !this.IsBlock);
    }

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    public Parse(parser: Parser, data: ParseData, state: State, scope: string, context: TagParseContext): INode {
        const index = state.Index;
        const tokenLength = this.Token.length;

        state.Seek(tokenLength + 1, false);
        let optionsString = state.ScanTo(']').trim();
        if (state.Next() != ']') {
            state.Seek(index, true);
            return null;
        }

        const options: Record<string, string> = {};
        if (optionsString.length > 0) {
            if (optionsString[0] == '=' && this.AllOptionsInMain && this.MainOption != null) {
                options[this.MainOption] = optionsString.substring(1);
            }
            else {
                if (optionsString[0] == '=') optionsString = this.MainOption + optionsString;
                const myregexp = /(?=\s|^)\s*([^ ]+?)=([^\s]*)(?=\s|$)(?!=)/img;
                let m = myregexp.exec(optionsString);
                while (m != null) {
                    const name = m[1].trim();
                    const value = m[2].trim();
                    options[name] = value;
                    m = myregexp.exec(optionsString);
                }
            }
        }

        if (this.IsNested) {
            let stack = 1;
            let text = '';
            while (!state.Done) {
                text += state.ScanTo('[');
                const tok = state.GetToken();
                if (tok.toLowerCase() == this.Token.toLowerCase()) stack++;
                if (tok.toLowerCase() == '/' + this.Token.toLowerCase() && state.Peek(tokenLength + 3).trim() == '[/' + this.Token.toLowerCase() + ']') stack--;
                if (stack == 0) {
                    state.Seek(this.Token.length + 3, false);
                    if (!this.Validate(options, text)) break;
                    return this.FormatResult(parser, data, state, scope, options, text);
                }

                text += state.Next();
            }

            state.Seek(index, true);
            console.log(2);
            return null;
        }
        else {
            const text = state.ScanTo('[/' + this.Token + ']', true);
            if (state.Peek(tokenLength + 3).trim() == '[/' + this.Token.toLowerCase() + ']' && this.Validate(options, text)) {
                state.Seek(this.Token.length + 3, false);
                return this.FormatResult(parser, data, state, scope, options, text);
            }
            else {
                state.Seek(index, true);
                console.log(3);
                return null;
            }
        }
    }

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    public Validate(options: Record<string, string>, text: string): boolean {
        return true;
    }

    public FormatResult(parser: Parser, data: ParseData, state: State, scope: string, options: Record<string, string>, text: string): INode {
        let before = '<' + this.Element;
        if (this.ElementClass != null) before += ' class="' + this.ElementClass + '"';
        before += '>';
        const after = '</' + this.Element + '>';
        const content = parser.ParseTags(data, text, scope, this.TagContext());
        const ret = new HtmlNode(before, content, after);
        ret.IsBlockNode = this.IsBlock;
        return ret;
    }


    // Extensions
    
    public WithScopes(...scopes : string[])
    {
        this.Scopes = scopes;
        return this;
    }

    public WithToken(token : string)
    {
        this.Token = token;
        return this;
    }

    public WithElement(element : string)
    {
        this.Element = element;
        return this;
    }

    public WithElementClass(elementClass : string)
    {
        this.ElementClass = elementClass;
        return this;
    }

    public WithBlock(isBlock : boolean)
    {
        this.IsBlock = isBlock;
        return this;
    }
}
